pkg/hive/hive.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hive
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+)
+
+const (
+	protocolName           = "hive"
+	initStreamName         = "hive_init"
+	initStreamVersion      = "1.0.0"
+	subscribeStreamName    = "hive_subscribe"
+	subscribeStreamVersion = "1.0.0"
+	peersStreamName        = "hive_peers"
+	peersStreamVersion     = "1.0.0"
+	//maxPeersCount     = 50
+)
+
+// messageTimeout is the maximal allowed time for a message to be read or written.
+var messageTimeout = 5 * time.Second
+
+type AddressBook interface {
+	// Add peer to knows peers (address book)
+	AddPeer(overlay, underlay []byte) error // todo: introduce bzz address if needed
+	ConnectedPeers() (peers []p2p.Peer)
+	UpdateDepth(peer p2p.Peer, depth uint8)
+}
+
+// SaturationTracker tracks weather the saturation has changed
+type SaturationTracker interface {
+	Subscribe() (c <-chan struct{}, unsubscribe func())
+	Depth() (uint8, error)
+}
+
+type Service struct {
+	streamer          p2p.Streamer
+	logger            logging.Logger
+	addressBook       AddressBook
+	saturationTracker SaturationTracker
+	quit              chan struct{}
+}
+
+type Options struct {
+	Streamer          p2p.Streamer
+	Logger            logging.Logger
+	PeerTracker       AddressBook
+	SaturationTracker SaturationTracker
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:          o.Streamer,
+		logger:            o.Logger,
+		saturationTracker: o.SaturationTracker,
+		addressBook:       o.PeerTracker,
+		quit:              make(chan struct{}),
+	}
+}
+
+func (s *Service) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name: protocolName,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    initStreamName,
```

Comment:
Hmmm. maybe we should talk about this on a meeting? It's a bit of a longer discussion. Basically td;dr; would be:
1. I talked with Viktor about having different messages here instead of using `Subscribe` and `Peers` messages on all places. This makes us differentiate how we handle them somehow, for example in `swarm` there is a state where we track done if it is a first response on `Subscribe` or not, which makes both sender and receiver handle these msg's differently in different context.
2. To avoid tracking this context, I have created the stream for `Init` (or handshake) phase, which is sending a `Subscribe` and receiving `Peers` msgs, and then additional handlers for regular`Subscribe` and `Peers`msg that are notifications
3. The above could also been accomplished by having a more different msg's. You can take a look at this PR:https://github.com/ethersphere/bee/pull/9. I had 2 more closed PR's about different approaches, you can check them, and let's discuss
4. One more point from me is that we shoulsd define where do we expect a response and where we do not. One example, if we want to know ,on `Init` , that both nodes subscribed and received `Peers` msg, we can use separate stream for this as well. 

These are just my 2 c, and it is not a complete explanation. Would you like to discuss it more on a meeting when you have time, would like to hear your opinion as well. Maybe, you, @janos  and myself could brainstorm it a bit?

Reply:
@pradovic I don't see any need for a handshake phase here (in fact I don't see one at all). All I see is two types of messages, I'm not sure what an extra stream brings as added value. It is not clear to me what is the semantics of a separated stream at this stage; do we infer that there is a stream per message? or that there's a stream per flow? in either case you should end up with two streams, not three.
Also, TBH having stuff like an `init` stream makes me feel we're already misusing this notion

Reply:
It is not actually, as the first time the Subscribe msg is sent it expects Peers msg as response, an in other cases it does not expect any response. I wanted to avoid having a state which tracks down is it the first time or not. If you had a chance to look at the PR, my solution to this was more msgs. This here was something that we talked on a daily last week and wanted to try out. But anyway, on the last meeting we concluded something completely different, so I will close this PR, as it does not makes sense now.